### ISSUE TITLE
JSON output format stability

### ISSUE THREAD

**Issue by tbzatek** [smartmontools/smartmontools#168]
Cc: @doug-gilbert

The commit 72491c57eec9e4d983ab9abd8910ec189612f8a2 has broken the JSON output format, comparing to the last public release of `smartmontools-7.3`.

Background: I've just started writing a `libblockdev` plugin that parses `smartctl` JSON output, for planned use in `udisks`. See e.g. related ticket #133.

Given that the `smartctl`'s JSON output is versioned (great feature btw.) I was expecting some degree of format stability. And just like commandline switches, return codes and console output (to some degree) might be considered a form of public ABI with some guarantees of stability, this naturally applies to JSON output as well. I.e. consumers parsing the output have some expectations and any disruption might break forward compatibility.

FWIW, I haven't found any JSON schema in the sources. I guess the guarantees need to be defined somewhere.

As for the mentioned commit, this may warrant bumping the `json_format_version`, we can modify our parser for both scenarios. Would be great to sort this out before the next public release.

**Comment by doug-gilbert** [comment]
That commit was on the first day (2022-01-05) that  _any_ JSON code was added to the SCSI subsystem in smartmontools (smt). I was immediately asked to prefix the new names with "scsi_" so they wouldn't conflict with names used by the ATA and NVMe subsystems. The release of 7.3 was on 2022-02-28, 7 weeks later. So I'm a bit surprised that you picked up that code on day 1 and not at the release. If you want to track this project with finer granularity than releases, perhaps you could look at the "git-svn-id" which shows the corresponding subversion revision number. The advantage of the svn ascending revision numbers (compared to git tags) is that it shows the time sequence. Incrementing the JSON version number needs doing, I agree, but when and what is the trigger? Probably the release of 7.4 in 3 months or so, would be a good point.
There is a lot more JSON output for SCSI code in smt since 2022-01-05, with more to come. I'm trying to stay close to the T10 naming and structure (which isn't always that self-consistent), plus consistent with my own tools (e.g. sg3_utils).
It is good to get feedback, because we are doing a lot of the JSON work "blind" and I'm learning as I go. Perhaps telling us when you are importing smt's JSON into a downstream tool would be useful (as you are doing now). I think a schema is some ways away (as it gives us enough rope to hang ourselves). For my work, I would point at the T10 documents as the "schemas".
Another issue that I have pondered is error processing and JSON. Since I have found no good solution my JSON output assumes no errors (IO or logical) and essentially stops building further JSON and dumps what has been gathered. The user gets a non-zero exit status and that is about it ...

**Comment by tbzatek** [comment]
You're right, there was acutally no change in JSON key names, comparing the 7.3 release tarball and the current svn snapshot. My apologies for this false alarm. What made me confused were these parts of the code:

https://github.com/smartmontools/smartmontools/blob/88534fdc9a4b90c6be6fb57a2316a03ab4a34b2a/smartmontools/scsiprint.cpp#L372-L385

https://github.com/smartmontools/smartmontools/blob/88534fdc9a4b90c6be6fb57a2316a03ab4a34b2a/smartmontools/scsiprint.cpp#L2980-L2995

Notice the `drive_trip` property and its section. Confronting the output from one of my SAS drives against the code (no JSON schema available so I had to read the sources to see what keys and values are available/valid), I came across different key naming and found a change to blame in the log. However overlooking the commit date...

**Comment by chrfranke** [comment]
Regarding JSON schema: I agree that this would make sense. Could you possibly recommend related tools and/or other projects which provide a schema?

Thanks for catching the inconsistency above. This was introduced in [r5286](https://www.smartmontools.org/changeset/5286) (and released in smartctl 7.3) on top the code from [r4815](https://www.smartmontools.org/changeset/4815) (smartctl 7.0).
We could now do any of:
1. Revert `scsi_temperature` back to `temperature` as it does not appear elsewhere (my vote).
2. Change `temperature` to `scsi_temperature` in line 2990 as wording as `drive_trip` seams to be SCSI/SAS specific.
3. Always provide `drive_trip` twice, as `scsi_temperature` and `temperature`.

@doug-gilbert: Thoughts?

**Comment by tbzatek** [comment]
One of the projects I'm following and which provides consistent JSON schema is `libnvme`: https://github.com/linux-nvme/libnvme/blob/master/doc/config-schema.json. It's a bit of manual work but it might possibly bring automated validation (e.g. as part of pre-release checks).

Generally, when developing a client that parses `smartctl --json` output, I find it somewhat difficult to discover what sections, keys and range of values are being offered. In my case the data type conversion chain goes like `smartctl` (cpp) -> console json output -> fork/exec -> `json-glib` (C) -> `libblockdev` (final C structures) with lots of potential for data conversion or parser issues. My point is to minimize the risk.

**Comment by doug-gilbert** [comment]
> 



> Regarding JSON schema: I agree that this would make sense. Could you possibly recommend related tools and/or other projects which provide a schema?
> 
> Thanks for catching the inconsistency above. This was introduced in [r5286](https://www.smartmontools.org/changeset/5286) (and released in smartctl 7.3) on top the code from [r4815](https://www.smartmontools.org/changeset/4815) (smartctl 7.0). We could now do any of:
> 
>     1. Revert `scsi_temperature` back to `temperature` as it does not appear elsewhere (my vote).
> 
>     2. Change `temperature` to `scsi_temperature` in line 2990 as wording as `drive_trip` seams to be SCSI/SAS specific.
> 
>     3. Always provide `drive_trip` twice, as `scsi_temperature` and `temperature`.
> 
> 
> @doug-gilbert: Thoughts?

Option 1 committed in revision 5451 . Will now look at associated pull request.

**Comment by doug-gilbert** [comment]
> One of the projects I'm following and which provides consistent JSON schema is `libnvme`: https://github.com/linux-nvme/libnvme/blob/master/doc/config-schema.json. It's a bit of manual work but it might possibly bring automated validation (e.g. as part of pre-release checks).

Not much meat in the above example. Twenty fields or less. smartmontools has hundreds when NVMe, ATA and SCSI are combined. I would like smartmontools to re-instate the capability to take in hexadecimal files rather than have a real device. Problem is those hexadecimal files will need some additional meta-data. Taking SCSI as an example, the hex could be from a standard INQUIRY response, READ CAPACITY, LOG SENSE, MODE SENSE and several others. This is simpler in sg3_utils as each utility tends to be at the level of each of those command responses. The 'inhex' folder in sg3_utils holds such hex files. Now when I add a new utility or add something to an existing utility (e.g. a new log page to sg_logs) I create a corresponding hex file in the inhex folder. In some cases that is from a real device (e.g. from a user '-HHH' report) or from my reading of a T10 spec.

I view hex as the pre-decoded form that will pick up some vendor extensions, and standardized stuff that hasn't yet been decoded by smartmontools. JSON can be viewed as the post-decoded form. In the JSON that I'm generating I'm trying to stay as close as possible to T10 field names, modulo snake_notation. But T10 are not that consistent, for example a lot of SCSI metadata uses a descriptor format which I translated to a JSON named array, however some pages have a "descriptor list" while other have a "descriptors_list" plus several other variations. Hard to change those once they are standardized.

Would it be useful to have those hex files that I have described above as JSON? Hex bytes are obviously an array of integers from 0 to 255. One advantage would be the ability to add meta-data which smartctl would need to make sense of it as input.

> Generally, when developing a client that parses `smartctl --json` output, I find it somewhat difficult to discover what sections, keys and range of values are being offered. In my case the data type conversion chain goes like `smartctl` (cpp) -> console json output -> fork/exec -> `json-glib` (C) -> `libblockdev` (final C structures) with lots of potential for data conversion or parser issues. My point is to minimize the risk.

And what does smartctl take as input in your test case?

**Comment by tbzatek** [comment]
True, maintaining any kind of JSON schema may be just too much effort.

I like your idea of having varied sample hex files for testing. It would certainly help with development of any kind of smartctl JSON output parser, covering variety of devices and their specific attributes.

> And what does smartctl take as input in your test case?

For now we only feed our json parser with captured output from smartctl on real devices. Nothing fancy but it tests the parser itself. Extending the tests by calling real smartctl and feeding it with sample hex data would certainly extend coverage on both projects. As our tests are part of our CI, regressions are caught quite early.
